Expand agent metric Policy Import Errors to count all policy changes

dlapcevic opened the pull request:
Change agent metric name from `policy_import_errors_total` to ``policy_change_total``. Now it counts all policy changes (Add, Update, Delete) based on outcome ("success" or "fail"). The metric can be used to show the percentage of success/failed network policy changes.

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

sayboras reviewed (commented):
The goal of this PR is making sense to me. 

I vaguely remember that we normally go with a 2-phase commit:
- Add new metrics and mark existing metrics as deprecated
- Remove deprecated metrics in the next release.

Another convention is to split the metric into 2 separate metrics (e.g. policy_change_success_total, policy_change_failure_total).

:thinking:

chancez reviewed (commented):
Generally I think this is reasonable, but I feel like we could potentially include more information in the metric about failures in particular.

I see a few places where we could have a "reason" that a policy change failed, for example: parsing errors. Is this something we should consider adding as a label? How many failure causes do we have for policy updates?

Additionally, it seems we could potentially have the source of the change (add/update/delete). Would any of these dimensions be useful for us?

Lastly, I noticed you added a whole new location to where these metrics are updated, in `pkg/k8s/watchers/cilium_network_policy.go` which previously did not touch the import errors metric. Aren't we double counting by doing it here and inside the `Handle` method?

sayboras commented on Documentation/operations/upgrade.rst:
```diff
@@ -312,6 +312,18 @@ Annotations:
+Deprecated Metrics
+~~~~~~~~~~~~~~~~~~
+
+* ``cilium_policy_import_errors_total`` is deprecated. Please use
```
thanks a lot :100:

sayboras replied:
nit:

Ideally, we try our best to leverage new metrics in dashboard/alert once they are introduced as these things are easierly overlooked.

https://github.com/cilium/cilium/blob/dcc458b5f423d5a22ab53b7888c69eb1d675d88b/examples/kubernetes/addons/prometheus/monitoring-example.yaml#L6031

https://github.com/cilium/cilium/blob/1626543f4164294d4afc576daf390c6f95a9de6a/examples/kubernetes/addons/prometheus/files/grafana-dashboards/cilium-dashboard.json#L5679

dlapcevic replied:
Done

sayboras reviewed (approved):
Can you help to rebase ? The diff looks good to me, only one minor comment on using new metrics in dashboard and monitor.

qmonnet commented on daemon/cmd/policy.go:
```diff
@@ -688,22 +688,26 @@ func (h *putPolicy) Handle(params PutPolicyParams) middleware.Responder {
 	var rules policyAPI.Rules
 	if err := json.Unmarshal([]byte(params.Policy), &rules); err != nil {
-		metrics.PolicyImportErrorsTotal.Inc()
+		metrics.PolicyImportErrorsTotal.Inc() // To be deprecated in Cilium 1.14.
```
As I understand, all of these should be instead:

```suggestion
		metrics.PolicyImportErrorsTotal.Inc() // Deprecated in Cilium 1.14, to be removed in 1.15.
```

Or did I get it wrong? We have branched out for 1.13 already, so these changes will be in 1.14 (as you say in the docs).

dlapcevic replied:
You are right. Done. Thanks.

qmonnet commented on pkg/metrics/metrics.go:
```diff
@@ -274,9 +274,15 @@ var (
 
-	// PolicyImportErrorsTotal is a count of failed policy imports
+	// PolicyImportErrorsTotal is a count of failed policy imports.
+	// This metric is going to be deprecated in Cilium 1.14 and removed in 1.15.
```
```suggestion
	// This metric is deprecated in Cilium 1.14 and is to be removed in 1.15.
```

dlapcevic replied:
Done

qmonnet reviewed (approved):
Looks OK to me, thanks

dlapcevic committed:
Expand agent metric Policy Import Errors to count all policy changes

Change agent metric name from policy_import_errors_total to policy_change_total. Now it counts all policy changes (Add, Update, Delete) based on outcome ("success" or "fail"). The metric can be used to show the percentage of success/failed network policy changes.

Signed-off-by: Dorde Lapcevic <dordel@google.com>

aditighag reviewed (commented):
Changes look fine to me. I was wondering how do we add labels for the failure cases specifically in future?   

>// Deprecated in Cilium 1.14, to be removed in 1.15.

Can you file an issue for this task?
